### ISSUE TITLE
VBCSCompiler crash with any custom WinRT events in VB

### ISSUE THREAD

**Issue by ljw1004** [dotnet/roslyn#6313]
REPRO: (this is the simplest example of a custom WinRT event that I could come up with)

(1) File>New>VB>Windows>Universal>BlankApp

(2) Paste the following code
```vb
Public NotInheritable Class C
    Implements IBackgroundTaskInstance

    Public Property InstanceId As Guid Implements IBackgroundTaskInstance.InstanceId
    Public Property Progress As UInteger Implements IBackgroundTaskInstance.Progress
    Public Property SuspendedCount As UInteger Implements IBackgroundTaskInstance.SuspendedCount
    Public Property Task As BackgroundTaskRegistration Implements IBackgroundTaskInstance.Task
    Public Property TriggerDetails As Object Implements IBackgroundTaskInstance.TriggerDetails
    Public Function GetDeferral() As BackgroundTaskDeferral Implements IBackgroundTaskInstance.GetDeferral
        Return Nothing
    End Function

    Private CanceledEvent As EventRegistrationTokenTable(Of BackgroundTaskCanceledEventHandler)

    Public Custom Event Canceled As BackgroundTaskCanceledEventHandler Implements IBackgroundTaskInstance.Canceled
        AddHandler(handler As BackgroundTaskCanceledEventHandler)
            Return EventRegistrationTokenTable(Of BackgroundTaskCanceledEventHandler).
                GetOrCreateEventRegistrationTokenTable(CanceledEvent).
                AddEventHandler(handler)
        End AddHandler
        RemoveHandler(token As EventRegistrationToken)
            CanceledEvent?.RemoveEventHandler(token)
        End RemoveHandler
        RaiseEvent(sender As IBackgroundTaskInstance, reason As BackgroundTaskCancellationReason)
            CanceledEvent?.InvocationList.Invoke(sender, reason)
        End RaiseEvent
    End Event

    Sub Trigger(reason As BackgroundTaskCancellationReason)
        RaiseEvent Canceled(Me, reason)
    End Sub

End Class
```

(3) Add the required Imports statement, and build the project.

WHAT I GET: During build, "VBCSCompiler has stopped working. Windows is sending the information to Microsoft..." and then "vbc.exe has stopped working..."

WHAT I EXPECT: It should build fine. Note that the corresponding C# code builds fine...

```cs
public sealed class C : IBackgroundTaskInstance
{
    public BackgroundTaskDeferral GetDeferral() => null;
    public Guid InstanceId { get; set; }
    public uint Progress { get; set; }
    public uint SuspendedCount { get; set; }
    public BackgroundTaskRegistration Task { get; set; }
    public object TriggerDetails { get; set; }

    private EventRegistrationTokenTable<BackgroundTaskCanceledEventHandler> CanceledEvent;

    public event BackgroundTaskCanceledEventHandler Canceled
    {
        add
        {
            return EventRegistrationTokenTable<BackgroundTaskCanceledEventHandler>
                .GetOrCreateEventRegistrationTokenTable(ref CanceledEvent)
                .AddEventHandler(value);
        }
        remove
        {
            CanceledEvent?.RemoveEventHandler(value);
        }
    }

    public void Trigger(BackgroundTaskCancellationReason reason)
    {
        CanceledEvent?.InvocationList.Invoke(this, reason);
    }

}
```

**Comment by TyOverby** [comment]
Callstack:

```
>	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol.SourceLocalSymbol.GetDeclaratorSyntax() Line 504	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.GetLocalDebugName(Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol local, Microsoft.CodeAnalysis.CodeGen.LocalDebugId localId) Line 1329	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.DefineLocal(Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol local, Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode syntaxNode) Line 1275	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitBlock(Microsoft.CodeAnalysis.VisualBasic.BoundBlock scope) Line 1227	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.EmitStatement(Microsoft.CodeAnalysis.VisualBasic.BoundStatement statement) Line 16	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.GenerateImpl() Line 163	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.CodeGen.CodeGenerator.Generate() Line 106	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.GenerateMethodBody(Microsoft.CodeAnalysis.VisualBasic.Emit.PEModuleBuilder moduleBuilder, Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol method, Integer methodOrdinal, Microsoft.CodeAnalysis.VisualBasic.BoundStatement block, System.Collections.Immutable.ImmutableArray(Of Microsoft.CodeAnalysis.CodeGen.LambdaDebugInfo) lambdaDebugInfo, System.Collections.Immutable.ImmutableArray(Of Microsoft.CodeAnalysis.CodeGen.ClosureDebugInfo) closureDebugInfo, Microsoft.CodeAnalysis.VisualBasic.StateMachineTypeSymbol stateMachineTypeOpt, Microsoft.CodeAnalysis.CodeGen.VariableSlotAllocator variableSlotAllocatorOpt, Microsoft.CodeAnalysis.CodeGen.DebugDocumentProvider debugDocumentProvider, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Boolean emittingPdb) Line 1510	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.LowerAndEmitMethod(Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol method, Integer methodOrdinal, Microsoft.CodeAnalysis.VisualBasic.BoundBlock block, Microsoft.CodeAnalysis.VisualBasic.Binder binderOpt, Microsoft.CodeAnalysis.VisualBasic.TypeCompilationState compilationState, Microsoft.CodeAnalysis.DiagnosticBag diagsForCurrentMethod, Microsoft.CodeAnalysis.VisualBasic.Binder.ProcessedFieldOrPropertyInitializers processedInitializers, Microsoft.CodeAnalysis.VisualBasic.SynthesizedSubmissionFields previousSubmissionFields, Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol constructorToInject, Integer delegateRelaxationIdDispenser) Line 1427	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.CompileMethod(Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol method, Integer methodOrdinal, Integer withEventPropertyIdDispenser, Integer delegateRelaxationIdDispenser, System.Predicate(Of Microsoft.CodeAnalysis.VisualBasic.Symbol) filter, Microsoft.CodeAnalysis.VisualBasic.TypeCompilationState compilationState, Microsoft.CodeAnalysis.VisualBasic.Binder.ProcessedFieldOrPropertyInitializers processedInitializers, Microsoft.CodeAnalysis.VisualBasic.Binder containingTypeBinder, Microsoft.CodeAnalysis.VisualBasic.SynthesizedSubmissionFields previousSubmissionFields, Microsoft.CodeAnalysis.VisualBasic.Symbols.MethodSymbol referencedConstructor) Line 1207	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.CompileNamedType(Microsoft.CodeAnalysis.VisualBasic.Symbols.NamedTypeSymbol containingType, System.Predicate(Of Microsoft.CodeAnalysis.VisualBasic.Symbol) filter) Line 653	Basic
 	Microsoft.CodeAnalysis.VisualBasic.dll!Microsoft.CodeAnalysis.VisualBasic.MethodCompiler.<closure>.<lambda0>() Line 495	Basic
 	Microsoft.CodeAnalysis.dll!Roslyn.Utilities.UICultureUtilities.WithCurrentUICulture.AnonymousMethod__0() Line 136	C#
```

**Comment by TyOverby** [comment]
Lines 499 and 504 look really suspect.  When running in debug mode, the assertion on 504 gets triggered, but that assert isn't present in the release compiler.

**Comment by TyOverby** [comment]
Closed in #6448
